core/src/main/java/io/atomix/core/transaction/impl/DefaultTransaction.java

```diff
@@ -127,7 +127,9 @@ void addParticipants(Collection<TransactionParticipant<?>> participants) {
             : transactionService.aborting(transactionId)
             .thenCompose(v -> rollback(participants))
             .thenApply(v -> CommitStatus.FAILURE));
-    return status.thenCompose(v -> transactionService.complete(transactionId).thenApply(u -> v));
+    return status.thenCompose(v -> transactionService.complete(transactionId)
+        .thenRun(() -> participants.forEach(TransactionParticipant::close))
```

Comment:
is it possible close may thow an exception and fail the future?

Reply:
I think it could indeed